### PR TITLE
Speed up processing of TimelinePanel updates on new events

### PR DESCRIPTION
**Pull request by richvdh**
Avoid twiddling state flags, so that we can avoid redundant re-renders.

### CODE REVIEW

File: src/components/structures/TimelinePanel.js [hunk 1]
```diff
@@ -267,13 +286,15 @@ var TimelinePanel = React.createClass({
...
-        this.setState({canForwardPaginate: true});
-
         if (!this.refs.messagePanel) return;
 
-        if (!this.refs.messagePanel.getScrollState().stuckAtBottom) return;
+        if (!this.refs.messagePanel.getScrollState().stuckAtBottom) {
+            // we won't load this event now, because we don't want to push any
+            // events off the other end of the timeline. But we need to note
+            // that we can now paginate.
+            this.setState({canForwardPaginate: true});
```
Comment (dbkr) [inline src/components/structures/TimelinePanel.js]: Does the act of changing this ever require a re-render? If not, would it be more appropriate to just make it a regular property on `this`?

Comment (richvdh) [inline src/components/structures/TimelinePanel.js]: The logic behind putting it in `state` was mostly for consistency with `canBackPaginate`. But yes, we could do that. Or we could ignore changes to `canForwardPaginate` in `shouldComponentUpdate`.

Comment (dbkr) [inline src/components/structures/TimelinePanel.js]: OK - hopefully doing either of those would reduce the number of re-renders further? Personally I'd make them regular properties and keep state for things that affect what's drawn to the screen.

**Comment by dbkr** [comment]
Comment aside, this fix is still valid either way so lgtm.

File: src/components/structures/TimelinePanel.js (hunk 1 above)
Comment (richvdh) [inline src/components/structures/TimelinePanel.js]: I need to return to this to fix https://github.com/vector-im/vector-web/issues/1427 anyway, so I'll go with it as-is for now.

**Pull request merged**